test/Microsoft.ML.Tests/Scenarios/Api/IntrospectiveTraining.cs

```diff
@@ -0,0 +1,118 @@
+                    NgramLength = 1,
+                    Column = new[] { new WordBagTransform.Column() { Name = "Tokenize", Source = new[] { "SentimentText" } } }
+                }, loader);
+
+                var lda = new LdaTransform(env, new LdaTransform.Arguments()
+                {
+                    NumTopic = 10,
+                    NumIterations = 3,
+                    NumThreads = 1,
+                    Column = new[] { new LdaTransform.Column { Source = "Tokenize", Name = "Features"}
+                    }
+                }, words);
+                var trainData = lda;
+
+                var cachedTrain = new CacheDataView(env, trainData, prefetch: null);
+                // Train the first predictor.
+                var linearTrainer = new LinearClassificationTrainer(env, new LinearClassificationTrainer.Arguments
+                {
+                    NumThreads = 1
+                });
+                var trainRoles = new RoleMappedData(cachedTrain, label: "Label", feature: "Features");
+                var linearPredictor = linearTrainer.Train(new Runtime.TrainContext(trainRoles));
+                VBuffer<float> weights = default;
+                linearPredictor.GetFeatureWeights(ref weights);
+
+                var topicSummary = lda.GetTopicSummary();
+                var treeTrainer = new FastTreeBinaryClassificationTrainer(env,
+                    new FastTreeBinaryClassificationTrainer.Arguments
+                    {
+                        NumTrees = 2
+                    }
+                    );
+                var ftPredictor = treeTrainer.Train(new Runtime.TrainContext(trainRoles));
+                FastTreeBinaryPredictor treePredictor;
+                if (ftPredictor is CalibratedPredictorBase calibrator)
+                    treePredictor = (FastTreeBinaryPredictor)calibrator.SubPredictor;
+                else
+                    treePredictor = (FastTreeBinaryPredictor)ftPredictor;
+                var featureNameCollection = FeatureNameCollection.Create(trainRoles.Schema);
+                foreach (var tree in treePredictor.GetTrees())
+                {
+                    var lteChild = tree.LteChild;
+                    var gteChild = tree.GtChild;
+                    // Get nodes.
+                    for (var i = 0; i < tree.NumNodes; i++)
+                    {
+                        var node = tree.GetNode(i, false, featureNameCollection);
+                        var gainValue = GetValue<double>(node.KeyValues, "GainValue");
+                        var splitGain = GetValue<double>(node.KeyValues, "SplitGain");
+                        var featureName = GetValue<string>(node.KeyValues, "SplitName");
+                        var previousLeafValue = GetValue<double>(node.KeyValues, "PreviousLeafValue");
+                        var threshold = GetValue<string>(node.KeyValues, "Threshold").Split(new[] { ' ' }, 2)[1];
+                        var nodeIndex = i;
+                    }
```

Comment:
You are missing categorical split nodes.

Reply:
I took this code from Tree Visualizer, which I expect contains all required code. Also I'm not sure how to get them, I think I call all exposed methods in this example.

Reply:
Tree visualizer contains code for categorical splits. Let me know if you need help.

---
In reply to: [209790640](https://github.com/dotnet/machinelearning/pull/653#discussion_r209790640) [](ancestors = 209790640)

Reply:
I can't find any word "categ" in that project, are you sure it's not part of GetNode or GetLeaf logic?

Reply:
FastTree.cs, line 3192. 

public INode GetNode(int nodeId, bool isLeaf, IEnumerable<string> featuresNames = null)
            {
                var keyValues = new Dictionary<string, object>();
                if (isLeaf)
                {
                    keyValues.Add(NodeKeys.LeafValue, _regTree.LeafValue(nodeId));
                }
                else
                {
                    if (featuresNames != null)
                    {
                        if (featuresNames is FeatureNameCollection features)
                        {
                            if (_regTree.CategoricalSplit[nodeId])
                            {
                                string featureList = string.Join(" OR \n",
                                    _regTree.CategoricalSplitFeatures[nodeId].Select(feature => features[feature]));

                                keyValues.Add(NodeKeys.SplitName, featureList);
                            }
                            else
                                keyValues.Add(NodeKeys.SplitName, features[_regTree.SplitFeature(nodeId)]);
                        }
                    }
                    keyValues.Add(NodeKeys.Threshold, string.Format("<= {0}", _regTree.RawThreshold(nodeId)));
                    if (_regTree.SplitGains != null)
                        keyValues.Add(NodeKeys.SplitGain, _regTree.SplitGains[nodeId]);
                    if (_regTree.GainPValues != null)
                        keyValues.Add(NodeKeys.GainValue, _regTree.GainPValues[nodeId]);
                    if (_regTree.PreviousLeafValues != null)
                        keyValues.Add(NodeKeys.PreviousLeafValue, _regTree.PreviousLeafValues[nodeId]);
                }


                return new TreeNode(keyValues);
            }

---
In reply to: [209791675](https://github.com/dotnet/machinelearning/pull/653#discussion_r209791675) [](ancestors = 209791675)

Reply:
Ok, and I'm getting SplitName from node (node, not leaf) in my code, which is represent categorical split features or just this exact feature.
Are you want me to do something with this information?

Reply:
This is not a comprehensive guide to FastTree internal structure. I think the code looked OK as it is: it just demonstrates that it is possible to inspect the model after training.


---
In reply to: [209795941](https://github.com/dotnet/machinelearning/pull/653#discussion_r209795941) [](ancestors = 209795941)

Reply:
Fair enough.

---
In reply to: [209990650](https://github.com/dotnet/machinelearning/pull/653#discussion_r209990650) [](ancestors = 209990650,209795941)